compiler/internal/analysis/defer.go

```diff
@@ -0,0 +1,108 @@
+//     are difficult to determine without [CFG] blocks, we treat all
+//     if-statements as if they are not terminating.
+//     That means there may be some false positives, since returns declared
+//     after a terminating branch will be marked as affected by a defer
+//     declared in that branch, when in reality they are not.
+//
+//   - Same as above but for else blocks, switch cases, and any branching.
+//
+//   - Loops (i.e. for-statements and for-range-statements) can cause return
+//     statements declared earlier in the loop to be affected by defers
+//     declared after it in the loop. We can't determine which branches in a
+//     loop may return to the start of the loop so we assume anywhere inside
+//     of a loop can return to the start of the loop.
+//     To handle this, all defers defined anywhere within a loop are assumed
+//     to affect any return also defined in that loop.
+//     We only need to track the top-level loop since nested loops will be
+//     superseded by the top-level loop.
+//
+//   - Labels and goto's are similar to loops in [CFG] blocks but without those
+//     blocks it's harder to determine which defers will affect which returns.
+//     To be safe, for any function with any blocking defers, returns, and
+//     goto's, all the returns are defaulted to blocking.
+//
+// [CFG]: https://en.wikipedia.org/wiki/Control-flow_graph
+type deferStmt struct {
+	isBlocking func(info *Info) (bool, bool)
+}
+
+// newBlockingDefer creates a new defer statement that is blocking or not.
+// If the defer is calling a js.Object method then the defer is non-blocking.
+// If the defers calling an interface method or function pointer in a var
+// then the defer is blocking.
+func newDefer(blocking bool) *deferStmt {
+	return &deferStmt{
+		isBlocking: func(info *Info) (bool, bool) {
+			return blocking, false
+		},
+	}
+}
+
+// newInstDefer creates a new defer statement for an instances of a method.
+// The instance is used to look up the blocking information later.
+func newInstDefer(inst typeparams.Instance) *deferStmt {
+	return &deferStmt{
+		isBlocking: func(info *Info) (bool, bool) {
+			return info.FuncInfo(inst).IsBlocking(), true
+		},
+	}
+}
+
+// newLitDefer creates a new defer statement for a function literal.
+// The literal is used to look up the blocking information later.
+func newLitDefer(lit *ast.FuncLit) *deferStmt {
+	return &deferStmt{
+		isBlocking: func(info *Info) (bool, bool) {
+			return info.FuncLitInfo(lit).IsBlocking(), true
+		},
+	}
+}
+
+// IsBlocking determines if the defer statement is blocking or not.
+//
+// The result will be cached for future calls since each return statement
+// will check the same defers (plus any new defers) of prior return statements.
+func (d *deferStmt) IsBlocking(info *Info) bool {
+	blocking, cacheResult := d.isBlocking(info)
+	if cacheResult {
+		d.isBlocking = func(info *Info) (bool, bool) {
+			return blocking, false
+		}
+	}
```

Comment:
Does caching actually provide a significant performance boost?

If not, I would avoid it, since it could be a cause for a difficult to debug inconsistency if at some future point functions could be declared blocked at a later processing stage.

Reply:
Good point, I'll make those more debuggable. The caching was only because I knew that each return would check all the defers the prior return checked plus some new ones (in general) and I was trying to reduce duplicate checks, but since those are just doing map lookups, it's probably not that much of a performance boost.